application/state_machine.go

```diff
@@ -0,0 +1,123 @@
+package application
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/ONSdigital/dp-dataset-api/models"
+	"github.com/ONSdigital/dp-dataset-api/store"
+)
+
+type State interface {
+	Enter(combinedVersionUpdate *models.Version, l *StateMachine, smDS *StateMachineDatasetAPI, ctx context.Context,
+		currentDataset *models.DatasetUpdate, // Called Dataset in Mongo
+		currentVersion *models.Version, // Called Instances in Mongo
+		versionUpdate *models.Version, // Next version, that is the new version
+		versionDetails VersionDetails) error
+	String() string
+}
+
+type StateMachine struct {
+	states      map[string]State
+	transitions map[string][]string
+	dataStore   store.DataStore
+	ctx         context.Context
+}
+
+type Transition struct {
+	Label                string
+	TargetState          State
+	AlllowedSourceStates []string
+}
+
+func castStateToState(state string) State {
+
+	switch s := state; s {
+	case "published":
+		return Published{}
+	case "associated":
+		return Associated{}
+	case "created":
+		return Created{}
+	case "completed":
+		return Completed{}
+	case "edition-confirmed":
+		return EditionConfirmed{}
+	case "detached":
+		return Detached{}
+	case "submitted":
+		return Submitted{}
+	case "failed":
+		return Failed{}
+	default:
+		return nil
+	}
+}
+
+func (sm *StateMachine) Transition(combinedVersionUpdate *models.Version, newState string, previousState string, smDS *StateMachineDatasetAPI, ctx context.Context,
+	currentDataset *models.DatasetUpdate, // Called Dataset in Mongo
+	currentVersion *models.Version, // Called Instances in Mongo
+	versionUpdate *models.Version, // Next version, that is the new version
+	versionDetails VersionDetails) error {
+
+	match := false
+	var nextState State
+
+	for state, transitions := range sm.transitions {
+		fmt.Println(state)
+		fmt.Println(transitions)
+
+		if state == newState {
+			fmt.Println("The states match")
+			for i := 0; i < len(transitions); i++ {
+				if previousState == transitions[i] {
+					match = true
+					nextState = castStateToState(newState)
```

Comment:
This call to the hard coded cast in the `castStateToState()` could be replaced with:

```suggestion
					nextState = sm.states[newState]
```
This would improve the maintainability of the code by preventing the need to extend the hard coded cast when changing the states.

Reply:
This has been amended